zipline/lib/_windowtemplate.pxi

```diff
@@ -0,0 +1,103 @@
+"""
+Template for AdjustedArray windowed iterators.
+
+This file is intended to be used by inserting it via a Cython include into a
+file that's define a type symbol named `ctype` and string constant named
+`dtype`.
+
+See Also
+--------
+zipline.lib._floatwindow
+zipline.lib._intwindow
+zipline.lib._datewindow
+"""
+from numpy cimport ndarray
+from numpy import asarray
+
+ctypedef ctype[:, :] databuffer
+
+
+cdef class AdjustedArrayWindow:
+    """
+    An iterator representing a moving view over an AdjustedArray.
+
+    Concrete subtypes should subclass this and provide a `data` attribute for
+    specific types.
+
+    This object stores a copy of the data from the AdjustedArray over which
+    it's iterating.  At each step in the iteration, it mutates its copy to
+    allow us to show different data when looking back over the array.
+
+    The arrays yielded by this iterator are always views over the underlying
+    data.
+    """
+    cdef:
+        # ctype must be defined by the file into which this is being copied.
+        databuffer data
+        object viewtype
+        readonly Py_ssize_t window_length
+        Py_ssize_t anchor, max_anchor, next_adj
+        dict adjustments
+        list adjustment_indices
+
+    def __cinit__(self,
+                  databuffer data not None,
+                  object viewtype not None,
+                  dict adjustments not None,
+                  Py_ssize_t offset,
+                  Py_ssize_t window_length):
+
+        self.data = data
+        self.viewtype = viewtype
+        self.adjustments = adjustments
+        self.adjustment_indices = sorted(adjustments, reverse=True)
+        self.window_length = window_length
+        self.anchor = window_length + offset
+        self.max_anchor = data.shape[0]
+
+        if len(self.adjustment_indices) > 0:
```

Comment:
this is duplicated in the `__next__`. Can we pull this out into a `next_adj` func?